In [41]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [42]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [43]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [4]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [44]:
print(df_market_data.shape)

# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaled_data =StandardScaler().fit_transform(df_market_data)
scaled_data


(41, 7)


array([[ 5.08529366e-01,  4.93193071e-01,  7.72200433e-01,
         2.35459633e-01, -6.74950963e-02, -3.55953481e-01,
        -2.51636882e-01],
       [ 1.85445894e-01,  9.34445040e-01,  5.58692121e-01,
        -5.43409317e-02, -2.73482725e-01, -1.15759474e-01,
        -1.99352110e-01],
       [ 2.17739616e-02, -7.06336853e-01, -2.16804207e-02,
        -6.10301536e-02,  8.00452481e-03, -5.50246924e-01,
        -2.82060506e-01],
       [-4.07643829e-02, -8.10928066e-01,  2.49457974e-01,
        -5.03879651e-02, -3.73164019e-01, -4.58258816e-01,
        -2.95546142e-01],
       [ 1.19303608e+00,  2.00095907e+00,  1.76061001e+00,
         5.45842065e-01, -2.91202870e-01, -4.99847761e-01,
        -2.70316950e-01],
       [ 8.91870708e-01,  1.32729453e+00,  8.00214184e-01,
        -5.71478992e-02,  7.78653106e-01, -1.88231917e-01,
        -2.25532605e-01],
       [ 1.13972400e-02,  2.57225091e+00,  1.10164693e+00,
        -4.90495415e-01, -9.31954023e-01,  3.87758986e-01,
        -1.8284399

In [45]:
df_market_data.columns

Index(['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'],
      dtype='object')

In [46]:
# Create a DataFrame with the scaled data
scaled_df =pd.DataFrame(scaled_data, columns=df_market_data.columns)
scaled_df

price_change_percentage_24h  price_change_percentage_7d  \
0                      0.508529                    0.493193   
1                      0.185446                    0.934445   
2                      0.021774                   -0.706337   
3                     -0.040764                   -0.810928   
4                      1.193036                    2.000959   
5                      0.891871                    1.327295   
6                      0.011397                    2.572251   
7                      0.102530                    1.508001   
8                      0.077497                    0.334297   
9                      0.448952                   -0.190684   
10                     0.331280                   -1.614844   
11                     0.034352                   -0.733026   
12                     0.155710                   -0.922491   
13                     0.262723                    1.792602   
14                     0.130050                   -0.041018   
15                    -0.151583                    0.708196   
16                    -0.923203                   -1.437359   
17                    -0.277543                   -0.385209   
18                    -0.255978                    1.840274   
19                     0.180851                   -0.704931   
20                     0.286546                   -0.326301   
21                     0.515453                    0.461843   
22                     0.051758                   -0.928381   
23                    -0.052032                   -0.457229   
24                    -0.217984                   -0.849381   
25                     0.061339                   -0.706669   
26                     0.259097                    0.249508   
27                     0.585089                   -0.994231   
28                    -0.127467                    0.929119   
29                    -1.612188                   -1.682027   
30                    -0.296940                    0.094763   
31                    -0.071312                   -0.229484   
32                    -4.981042                   -0.045178   
33                    -0.125168                    0.580730   
34                    -1.428574                   -0.025510   
35                     1.919812                    0.370447   
36                     1.045530                   -0.618328   
37                    -0.409044                   -0.906963   
38                     0.414711                    0.414044   
39                     0.078038                   -0.687745   
40                     1.217453                   -0.607714   

    price_change_percentage_14d  price_change_percentage_30d  \
0                      0.772200                     0.235460   
1                      0.558692                    -0.054341   
2                     -0.021680                    -0.061030   
3                      0.249458                    -0.050388   
4                      1.760610                     0.545842   
5                      0.800214                    -0.057148   
6                      1.101647                    -0.490495   
7                      0.648885                     0.328959   
8                      0.858520                    -0.012646   
9                     -0.248043                     0.051634   
10                    -1.054521                    -0.729931   
11                    -0.023140                    -0.065775   
12                     0.115024                    -0.237488   
13                     2.202665                     1.437842   
14                     0.147155                    -0.543776   
15                     0.258012                    -0.602296   
16                    -0.629963                    -0.460558   
17                    -0.153243                    -0.371816   
18                     0.643565                     0.116538   
19                    -0.001816                    -0.143237   
20           

In [47]:
#  Copy the crypto names from the original data
crypto_names = df_market_data.index
crypto_names

Index(['bitcoin', 'ethereum', 'tether', 'ripple', 'bitcoin-cash',
       'binancecoin', 'chainlink', 'cardano', 'litecoin', 'bitcoin-cash-sv',
       'crypto-com-chain', 'usd-coin', 'eos', 'monero', 'tron', 'tezos', 'okb',
       'stellar', 'cosmos', 'cdai', 'neo', 'wrapped-bitcoin', 'leo-token',
       'huobi-token', 'nem', 'binance-usd', 'iota', 'vechain', 'zcash',
       'theta-token', 'dash', 'ethereum-classic', 'ethlend', 'maker', 'havven',
       'omisego', 'celsius-degree-token', 'ontology', 'ftx-token', 'true-usd',
       'digibyte'],
      dtype='object', name='coin_id')

In [48]:
scaled_df['coin_id']=crypto_names
# Set the coinid column as index make new column
scaled_df.set_index('coin_id', drop=True, inplace=True)
# Display sample data
scaled_df

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

---

### Find the Best Value for k Using the Original Data.

In [49]:
# Create a list with the number of k-values from 1 to 11
k_values= list(range(1,12))
k_values

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [50]:
# Create an empty list to store the inertia values
inertia_values = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
for k in k_values:
    print(k)
    # 1. Create a KMeans model using the loop counter for the n_clusters
    k_model = KMeans(n_clusters=k, random_state=1)
    # 2. Fit the model to the data using `df_market_data_scaled`
    k_model.fit(scaled_df)
    # 3. Append the model.inertia_ to the inertia list
    inertia_values.append(k_model.inertia_)

1
2
3
4


c:\Users\19802\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1037: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  "KMeans is known to have a memory leak on Windows "


5
6
7
8
9
10
11


In [51]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k_values, "inertia":inertia_values}

# Create a DataFrame with the data to plot the Elbow curve
elbow_df =pd.DataFrame(elbow_data)
elbow_df

k     inertia
0    1  287.000000
1    2  195.820218
2    3  123.190482
3    4   79.022435
4    5   65.405923
5    6   52.933559
6    7   47.983124
7    8   37.288187
8    9   33.061685
9   10   28.779752
10  11   25.248499

In [13]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_df.hvplot.line(
    x="k",
    y= "inertia",
    title= "Elbow Curve 1",
    xticks =k
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** the best value of k=4 which is 79.022 

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [52]:
# Initialize the K-Means model using the best value for k
model =  KMeans(n_clusters = 4,random_state = 1)
model

KMeans(n_clusters=4, random_state=1)

In [53]:
# Fit the K-Means model using the scaled data
model.fit(scaled_df)

KMeans(n_clusters=4, random_state=1)

In [54]:
# Predict the clusters to group the cryptocurrencies using the scaled data
k_4= model.predict(scaled_df)

# Print the resulting array of cluster values.
k_4

array([3, 3, 1, 1, 3, 3, 3, 3, 3, 1, 1, 1, 1, 3, 1, 3, 1, 1, 3, 1, 1, 3,
       1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 0, 3, 1, 1, 2, 1, 1, 1, 1])

In [55]:
# Create a copy of the DataFrame
copy_scaled_df= scaled_df.copy()
copy_scaled_df

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

In [56]:
# Add a new column to the DataFrame with the predicted clusters
copy_scaled_df['predicted_clusters']= k_4
# Display sample data
copy_scaled_df.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  predicted_clusters  
coin_id                                                       
bitcoin                        -0.251637                   3  
ethereum                       -0.199352                   3  
tether                         -0.282061                   1  
ripple                         -0.295546                   1  
bitcoin-cash                   -0.270317                   3

In [57]:
# Create a scatter plot using hvPlot by setting 
# Plot the clusters
copy_scaled_df.hvplot.scatter(
    # `x="price_change_percentage_24h"
    x="price_change_percentage_24h",
    # y="price_change_percentage_7d"`.
    y="price_change_percentage_7d",
    by="predicted_clusters",
    hover_cols ="coin_id",
    title= "Predicted clusters"
)

:NdOverlay   [predicted_clusters]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [58]:
# Create a PCA model instance and set `n_components=3`.
pca= PCA(n_components=3)
pca

PCA(n_components=3)

In [59]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
clusters_pca = pca.fit_transform(copy_scaled_df)
# View the first five rows of the DataFrame. 
clusters_pca[:5]


array([[-1.58513942,  0.13439444,  0.545621  ],
       [-1.32258066,  0.05449393,  1.11570854],
       [ 0.26633536, -0.53640937, -0.79694588],
       [ 0.25722513, -0.59201286, -0.6465287 ],
       [-2.95224985,  0.42237155,  1.26789353]])

In [60]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.37005408, 0.32322221, 0.19115222])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.8844285111826466

In [23]:
sum(pca.explained_variance_ratio_)

0.8844285111826466

In [61]:
# Create a new DataFrame with the PCA data.# Creating a DataFrame with the PCA data
clusters_pca_df= pd.DataFrame(
    clusters_pca,
    columns =["PC1","PC2","PC3"])
clusters_pca_df
# Copy the crypto names from the original data
clusters_pca_df['coin_id']= list(copy_scaled_df.index)

# Set the coinid column as index
clusters_pca_df= clusters_pca_df.set_index('coin_id')

# Display sample data
clusters_pca_df

PC1       PC2       PC3
coin_id                                           
bitcoin              -1.585139  0.134394  0.545621
ethereum             -1.322581  0.054494  1.115709
tether                0.266335 -0.536409 -0.796946
ripple                0.257225 -0.592013 -0.646529
bitcoin-cash         -2.952250  0.422372  1.267894
binancecoin          -1.975484  0.528219  0.670145
chainlink            -1.671906  0.178423  2.718598
cardano              -1.539740  0.324517  1.497226
litecoin             -1.347533 -0.132837  0.791063
bitcoin-cash-sv       0.040698 -0.802083 -0.465562
crypto-com-chain      1.344822 -1.108679 -1.076912
usd-coin              0.272121 -0.545849 -0.814012
eos                   0.320673 -0.918019 -0.637918
monero               -2.838740  1.895142  0.848863
tron                  0.143758 -0.442297 -0.274682
tezos                -0.862155 -0.757488  1.490647
okb                   1.378653 -0.776673 -0.754383
stellar               0.532658 -0.876910 -0.120776
cosmos               -1.539433  0.352390  1.713933
cdai                  0.208186 -0.588555 -0.830587
neo                   0.887872 -0.947174 -0.588142
wrapped-bitcoin      -1.572600  0.121709  0.530894
leo-token             0.697532 -0.826190 -0.993080
huobi-token           0.263251 -0.527831 -0.530039
nem                   0.408806  0.707042 -1.330139
binance-usd           0.246128 -0.540957 -0.809982
iota                  0.186495 -0.967371  0.003609
vechain               1.528493 -1.668876 -0.750235
zcash                -1.347135  0.106093  1.200966
theta-token           2.056222  2.034190 -1.525531
dash                  0.222043 -0.817068  0.122347
ethereum-classic      0.228151 -0.739212 -0.299732
ethlend               7.458259  4.338598  3.751968
maker                -0.991935 -0.079763  0.741439
havven                2.271195 -0.622938  0.875803
omisego               0.234856  0.223448 -1.390103
celsius-degree-token -1.949062  7.705390 -3.110327
ontology              1.510753 -1.801827 -0.193429
ftx-token            -0.092618 -0.521653 -0.153336
true-usd              0.229197 -0.538204 -0.805785
digibyte              0.393926 -0.449545 -0.988556

---

### Find the Best Value for k Using the PCA Data

In [62]:
# Create a list with the number of k-values from 1 to 11
k= list(range(1,12))
k

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [63]:
# Create an empty list to store the inertia values
inertia2=[]
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
for i in k: 
    print(i)
    # 1. Create a KMeans model using the loop counter for the n_clusters
    k_model=KMeans(n_clusters=i, random_state=1)
    # 2. Fit the model to the data using `df_market_data_pca`
    k_model.fit(clusters_pca_df)
    # 3. Append the model.inertia_ to the inertia list
    inertia2.append(k_model.inertia_)


1
2
3
4


c:\Users\19802\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1037: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  "KMeans is known to have a memory leak on Windows "


5
6
7
8
9
10
11


In [64]:
# Create a dictionary with the data to plot the Elbow curve
elbow_dic ={"k": k, "inertia":inertia2 }
# Create a DataFrame with the data to plot the Elbow curve
elbow_df2 = pd.DataFrame(elbow_dic)
elbow_df2

k     inertia
0    1  287.007838
1    2  196.268358
2    3  111.972027
3    4   43.205879
4    5   33.042435
5    6   24.347640
6    7   19.777320
7    8   15.868897
8    9   12.351058
9   10   10.169351
10  11    8.095958

In [65]:
# Plot a line chart with all the inertia values computed with 
elbow_df2.hvplot.line(
    x= "k",
    y="inertia",
    title= "Elbow line Curve 2",
    xticks=k
)
# the different values of k to visually identify the optimal value for k.

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No 

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [66]:
# Initialize the K-Means model using the best value for k
model_best= KMeans(n_clusters =4, random_state=1)
model_best

KMeans(n_clusters=4, random_state=1)

In [67]:
# Fit the K-Means model using the PCA data
model.fit(clusters_pca_df)

KMeans(n_clusters=4, random_state=1)

In [31]:
# Predict the clusters to group the cryptocurrencies using the PCA data
K_4= model.predict(clusters_pca_df)
# Print the resulting array of cluster values.
k_4

array([3, 3, 1, 1, 3, 3, 3, 3, 3, 1, 1, 1, 1, 3, 1, 3, 1, 1, 3, 1, 1, 3,
       1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 0, 3, 1, 1, 2, 1, 1, 1, 1])

In [68]:
# Create a copy of the DataFrame with the PCA data
copy_clusters_pca_df= clusters_pca_df.copy()
copy_clusters_pca_df
# Add a new column to the DataFrame with the predicted clusters
copy_clusters_pca_df['predicted_clusters']= k_4

# Display sample data
copy_clusters_pca_df

PC1       PC2       PC3  predicted_clusters
coin_id                                                               
bitcoin              -1.585139  0.134394  0.545621                   3
ethereum             -1.322581  0.054494  1.115709                   3
tether                0.266335 -0.536409 -0.796946                   1
ripple                0.257225 -0.592013 -0.646529                   1
bitcoin-cash         -2.952250  0.422372  1.267894                   3
binancecoin          -1.975484  0.528219  0.670145                   3
chainlink            -1.671906  0.178423  2.718598                   3
cardano              -1.539740  0.324517  1.497226                   3
litecoin             -1.347533 -0.132837  0.791063                   3
bitcoin-cash-sv       0.040698 -0.802083 -0.465562                   1
crypto-com-chain      1.344822 -1.108679 -1.076912                   1
usd-coin              0.272121 -0.545849 -0.814012                   1
eos                   0.320673 -0.918019 -0.637918                   1
monero               -2.838740  1.895142  0.848863                   3
tron                  0.143758 -0.442297 -0.274682                   1
tezos                -0.862155 -0.757488  1.490647                   3
okb                   1.378653 -0.776673 -0.754383                   1
stellar               0.532658 -0.876910 -0.120776                   1
cosmos               -1.539433  0.352390  1.713933                   3
cdai                  0.208186 -0.588555 -0.830587                   1
neo                   0.887872 -0.947174 -0.588142                   1
wrapped-bitcoin      -1.572600  0.121709  0.530894                   3
leo-token             0.697532 -0.826190 -0.993080                   1
huobi-token           0.263251 -0.527831 -0.530039                   1
nem                   0.408806  0.707042 -1.330139                   1
binance-usd           0.246128 -0.540957 -0.809982                   1
iota                  0.186495 -0.967371  0.003609                   1
vechain               1.528493 -1.668876 -0.750235                   1
zcash                -1.347135  0.106093  1.200966                   3
theta-token           2.056222  2.034190 -1.525531                   1
dash                  0.222043 -0.817068  0.122347                   1
ethereum-classic      0.228151 -0.739212 -0.299732                   1
ethlend               7.458259  4.338598  3.751968                   0
maker                -0.991935 -0.079763  0.741439                   3
havven                2.271195 -0.622938  0.875803                   1
omisego               0.234856  0.223448 -1.390103                   1
celsius-degree-token -1.949062  7.705390 -3.110327                   2
ontology              1.510753 -1.801827 -0.193429                   1
ftx-token            -0.092618 -0.521653 -0.153336                   1
true-usd              0.229197 -0.538204 -0.805785                   1
digibyte              0.393926 -0.449545 -0.988556                   1

In [70]:
# Create a scatter plot using hvPlot by setting # `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
copy_clusters_pca_df.hvplot.scatter(
   x="PC1",
   y="PC2",
   by= "predicted_clusters",
   hover_cols="coin_id",
   title= "The Cryptocurrency Represented By Each PCA Data"
)

:NdOverlay   [predicted_clusters]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [69]:
# Composite plot to contrast the Elbow curves
elbow_df.hvplot.line(x="k",y= "inertia",title= "Elbow Curve 1", xticks= k)+ elbow_df2.hvplot.line(x="k",y="inertia",title="Elbow Line Curve 2")

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [71]:
# Composite plot to contrast the clusters 
# YOUR CODE HERE!
copy_scaled_df.hvplot.scatter(x="price_change_percentage_24h", y= "price_change_percentage_7d",by="predicted_clusters",hover_cols = "coin_id",title= "Predicted clusters")+copy_clusters_pca_df.hvplot.scatter(x="PC1", y="PC2", hover_cols="coin_id",
title= "The Cryptocurrency Represented By Each PCA Data")

:Layout
   .NdOverlay.I :NdOverlay   [predicted_clusters]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .Scatter.I   :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:**  The impact of using PCA features data resule the data point seem to be the same cluster. 